File: specs/ts/line-multi-series.ts
import { Spec } from '@uwdata/mosaic-spec';

export const spec : Spec = {
  "meta": {
    "title": "Line Multi-Series",
    "description": "This line chart shows the unemployment rate of various U.S. metro divisions from 2000 through 2013. On hover, the closest data point to the pointer and its associated series is highlighted. Highlighting of series is performed using `nearestX` and `highlight` interactors. Point and text annotations instead use the mark `select` filter option.\n",
    "credit": "Adapted from a [D3 example](https://observablehq.com/@d3/multi-line-chart/2). Data from the [Bureau of Labor Statistics](https://www.bls.gov/).\n"
  },
  "data": {
    "bls_unemp": {
      "file": "data/bls-metro-unemployment.parquet"
    }
  },
  "plot": [
    {
      "mark": "ruleY",
      "data": [
        0
      ]
    },
    {
      "mark": "lineY",
      "data": {
        "from": "bls_unemp",
        "optimize": false
      },
      "x": "date",
      "y": "unemployment",
      "z": "division",
      "stroke": "steelblue",
      "curve": "monotone-x",
      "mixBlendMode": "multiply"
    },
    {
      "select": "nearestX",
      "channels": [
        "z"
      ],
      "as": "$curr"
    },
    {
      "select": "highlight",
      "by": "$curr"
    },
    {
      "mark": "dot",
      "data": {
        "from": "bls_unemp"
      },
      "x": "date",
      "y": "unemployment",
      "z": "division",
      "r": 2,
      "fill": "currentColor",
      "select": "nearestX"
    },
    {
      "mark": "text",
      "data": {
        "from": "bls_unemp"
      },
      "x": "date",
      "y": "unemployment",
      "text": "division",
      "fill": "currentColor",
      "dy": -8,
      "select": "nearestX"
    }
  ],
  "marginLeft": 24,
  "xLabel": null,
  "xTicks": 10,
  "yLabel": "Unemployment (%)",
  "yGrid": true,
  "style": "overflow: visible;",
  "width": 680
};

File: specs/python/line-multi-series.py
import vgplot as vg

bls_unemp = vg.parquet("data/bls-metro-unemployment.parquet")

curr = vg.selection.intersect()

view = vg.plot(
    vg.rule_y(data=[0]),
    vg.line_y(
        data="bls_unemp",
        optimize=False,
        x="date",
        y="unemployment",
        z="division",
        stroke="steelblue",
        stroke_opacity=0.9,
        curve="monotone-x",
    ),
    vg.nearest_x(channels=["z"], bind=curr),
    vg.highlight(by=curr),
    vg.dot(
        bls_unemp,
        x="date",
        y="unemployment",
        z="division",
        r=2,
        fill="currentColor",
        select="nearestX",
    ),
    vg.text(
        bls_unemp,
        x="date",
        y="unemployment",
        text="division",
        fill="currentColor",
        dy=-8,
        select="nearestX",
    ),
    vg.margin_left(24),
    vg.x_label(None),
    vg.x_ticks(10),
    vg.y_label("Unemployment (%)"),
    vg.y_grid(True),
    vg.style("overflow: visible;"),
    vg.width(680),
)

spec = vg.spec()
